pkg/vulnsrc/openeuler/openeuler.go

```diff
@@ -0,0 +1,267 @@
+			arches := lo.Uniq(pkg.Archs)
+			sort.Strings(arches)
+			advisory := types.Advisory{
+				FixedVersion: pkg.FixedVersion,
+				Arches:       arches,
+			}
+			// Don't put the same data source multiple times.
+			if _, ok := uniqOSVers[pkg.OSVer]; !ok {
+				uniqOSVers[pkg.OSVer] = struct{}{}
+				if err := vs.dbc.PutDataSource(tx, pkg.OSVer, source); err != nil {
+					return xerrors.Errorf("failed to put data source: %w", err)
+				}
+			}
+
+			if err := vs.dbc.PutAdvisoryDetail(tx, cvrf.Tracking.ID, pkg.Name,
+				[]string{pkg.OSVer}, advisory); err != nil {
+				return xerrors.Errorf("unable to save %s CVRF: %w", pkg.OSVer, err)
+			}
+		}
+
+		var references []string
+		for _, ref := range cvrf.References {
+			references = append(references, ref.URL)
+		}
+
+		severity := types.SeverityUnknown
+		for _, cvuln := range cvrf.Vulnerabilities {
+			for _, threat := range cvuln.Threats {
+				sev := severityFromThreat(threat.Severity)
+				if severity < sev {
+					severity = sev
+				}
+			}
+		}
+
+		vuln := types.VulnerabilityDetail{
+			References:  references,
+			Title:       cvrf.Title,
+			Description: getDetail(cvrf.Notes),
+			Severity:    severity,
+		}
+		if err := vs.dbc.PutVulnerabilityDetail(tx, cvrf.Tracking.ID, source.ID, vuln); err != nil {
+			return xerrors.Errorf("failed to save openEuler CVRF vulnerability: %w", err)
+		}
+
+		// for optimization
+		if err := vs.dbc.PutVulnerabilityID(tx, cvrf.Tracking.ID); err != nil {
+			return xerrors.Errorf("failed to save the vulnerability ID: %w", err)
+		}
+	}
+	return nil
+}
+
+func getAffectedPackages(productTree ProductTree) []Package {
+	var pkgs = make(map[string]Package) // pkgID => Package
+	for _, branch := range productTree.Branches {
+		// `src` pkgs are not installed in openEuler.
+		if branch.Type != "Package Arch" || branch.Name == "src" || branch.Name == "" {
+			continue
+		}
+		for _, production := range branch.Productions {
+			osVer := getOSVersion(production.CPE)
+			if osVer == "" {
+				log.Printf("unable to parse OS version: %s", production.CPE)
+				continue
+			}
+
+			// e.g., `ignition-debuginfo-2.14.0-2` or `perf-5.10.0-153.48.0.126.oe2203sp2.aarch64.rpm`
+			pkgName, pkgVersion := parseProduction(production)
+			if pkgName == "" || pkgVersion == "" {
+				log.Printf("unable to parse Production: %s", production)
+				continue
+			}
+
+			pkg := Package{
+				Name:         pkgName,
+				FixedVersion: pkgVersion,
+				Archs:        []string{branch.Name},
+				OSVer:        osVer,
+			}
+
+			id := fmt.Sprintf("%s-%s-%s", pkgName, pkgVersion, osVer)
+			if p, ok := pkgs[id]; ok {
+				pkg.Archs = append(pkg.Archs, p.Archs...)
+			}
+
+			pkgs[id] = pkg
+		}
+	}
+	return lo.Values(pkgs)
+}
+
+func getOSVersion(cpe string) string {
+	// e.g. cpe:/a:openEuler:openEuler:22.03-LTS-SP3
+	parts := strings.Split(cpe, ":")
+	if len(parts) != 5 || parts[2] != "openEuler" {
+		return ""
+	}
+	version := parts[4]
+	// e.g. 23.09, 22.03-LTS, 22.03-LTS-SP3
```

Comment:
I tend to preserve those versions, since they have different lifecycles (https://www.openeuler.org/en/other/lifecycle/) and different fixed versions. 

For example:
The fixed version of `kernel` in openEuler-22.03-LTS-SP3 is `5.10.0-217.0.0.120`, but that in 22.03-LTS-SP4 is `kernel-5.10.0-217.0.0.116`

see https://repo.openeuler.org/security/data/cvrf/2024/cvrf-openEuler-SA-2024-1792.xml and https://repo.openeuler.org/security/data/cvrf/2024/cvrf-openEuler-SA-2024-1795.xml